crates/oci-distribution/src/reference.rs

```diff
@@ -0,0 +1,82 @@
+use std::convert::{Into, TryFrom};
+
+// currently, the library only accepts modules tagged in the following structure:
+// <registry>/<repository>:<tag>
+// for example: webassembly.azurecr.io/hello:v1
+#[derive(Clone)]
+pub struct Reference {
+    whole: String,
+    slash: usize,
+    colon: usize,
+}
+
+impl Reference {
+    pub fn whole(&self) -> &str {
+        &self.whole
+    }
+
+    pub fn registry(&self) -> &str {
+        &self.whole[..self.slash]
+    }
+
+    pub fn repository(&self) -> &str {
+        &self.whole[self.slash + 1..self.colon]
+    }
+
+    pub fn tag(&self) -> &str {
+        &self.whole[self.colon + 1..]
+    }
+
+    pub fn to_v2_manifest_url(&self) -> String {
+        format!(
+            "https://{}/v2/{}/manifests/{}",
+            self.registry(),
+            self.repository(),
+            self.tag()
+        )
+    }
+}
+
+impl TryFrom<String> for Reference {
+    type Error = ();
```

Comment:
Do we want to give this a more descriptive name?

```rust
pub struct ReferenceParseError {
    _priv: ()
}
```

Reply:
This is moved over from Wok. I rewrote it today. I don't know the history behind it. But it was far more helpful with some actual errors instead of returning `()` on failure. I found it very confusing (see earlier thing about this not returning an error)